Add Python 3.10 support

micahellison opened the pull request:
<!--
Thank you for wanting to contribute!

Please fill out this description, and the checklist below.

Here are some key points to include in your description:
- What is this new code intended to do?
- Are there any related issues?
- What is the motivation for this change?
- What is an example of usage, or changes to config files? (if applicable)
-->
Fixes #1270, if all the tests pass on all platforms.

### Checklist

- [x] I have read the [contributing doc](https://github.com/jrnl-org/jrnl/blob/develop/CONTRIBUTING.md).
- [x] I have included a link to the relevant issue number.
- [ ] I have tested this code locally.
- [x] I have checked to ensure there aren't other open [pull requests](../pulls)
  for the same issue.
- [ ] I have written new tests for these changes, as needed.
- [x] All tests pass.
<!--
NOTE: Your PR may not be reviewed if there are any failing tests. You can run
tests locally with `make test` (see the contributing doc if you need help with
`make`), or use our automated tests after you submit your PR.
-->

micahellison committed:
Add Python 3.10 support

micahellison committed:
Change Python 3.10 to 3.10-dev in CI

micahellison committed:
Run poetry update

wren commented on .github/workflows/testing.yaml:
```diff
@@ -25,11 +25,8 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
```
```suggestion
    strategy:
      fail-fast: false
```

wren commented on .github/workflows/testing.yaml:
```diff
@@ -25,11 +25,8 @@ jobs:
     runs-on: ${{ matrix.os }}
```
```suggestion
    runs-on: ${{ matrix.os }}
    continue-on-error: ${{ matrix.python-version == '3.10-dev' }}
```

micahellison committed:
Turn off fail-fast to troubleshoot

Co-authored-by: Jonathan Wren <jonathan@nowandwren.com>

micahellison committed:
Add continue-on-error for 3.10

Co-authored-by: Jonathan Wren <jonathan@nowandwren.com>

musicinmybrain commented:
I managed to reproduce the CI failure in an Ubuntu VM. I also found a report at https://bugzilla.redhat.com/show_bug.cgi?id=1970361, the corresponding upstream bug (https://github.com/python-poetry/poetry/issues/3897), and the PR that fixed the bug: https://github.com/python-poetry/poetry/pull/3895. The PR was merged into master, but it seems it has not been backported to the `1.1` branch, so the bug is still present in the latest release of `poetry`.

wren commented:
In that case, we could probably install poetry 1.2.0a1 to get through this. I tested locally, and that seems to fix the error. 

I'm hesitant to use an alpha build for our entire pipeline, though, but I'd be okay with doing it just for the python 3.10-dev tests.

wren committed:
use prerelease poetry for prerelease python

wren committed:
fix syntax error

wren committed:
rename steps to make more sense

wren committed:
remove dev dependency that breaks with python 3.10

wren committed:
update tests

wren committed:
Change test dependencies for minimal breakage on python 3.10

- loosen the test dep requirements (so we get warned about these
  problems sooner)
- add new extras group to provide minimal deps required for testing (we
  don't need to run static analysis again on every version)

wren committed:
change how we check if deps are installed so test don't run wild

wren committed:
add setuptools due to poetry bug, clean up other steps

wren committed:
update lock file

wren committed:
test

wren commented:
Well, this isn't great. I ran into python-poetry/poetry#3153 while updating our dependencies. I think we might have to install poetry v1.2.0a1 for all of our tests, or I'm afraid they might not run anymore. I did it and now all of the tests are passing.

Thoughts on using an alpha build for our entire pipeline?

wren committed:
Revert "test"

This reverts commit 31e538300ef7f2ab3ff84da7fe03f156e4714e06.

micahellison commented:
Turns out these GitHub Actions failures were a fluke - we reverted and ran again and everything passes.

wren reviewed (approved):
lgtm 👩‍⚕️
